README.md

# SpotiTube
### A Converter That Will Convert Your Spotify To YT with LavaLink.
[![Discord](https://img.shields.io/discord/469387242767646730?style=flat-square&logo=discord&logoColor=white)](https://discordapp.com/invite/GuGcuwdYwg)
![Github Stars](https://img.shields.io/github/stars/BotGrace/SpotiTube?style=flat-square)
![GitHub issues](https://img.shields.io/github/issues-raw/BotGrace/SpotiTube?style=flat-square)
[![npm](https://img.shields.io/npm/v/spotitube?style=flat-square&maxAge=3600)](https://www.npmjs.com/package/spotitube)
[![npm downloads](https://img.shields.io/npm/dt/spotitube?style=flat-square&maxAge=3600)](https://www.npmjs.com/package/spotitube)
![Snyk Vulnerabilities for npm package](https://img.shields.io/snyk/vulnerabilities/npm/spotitube?style=flat-square&maxAge=3600) 
![NPM](https://img.shields.io/npm/l/spotitube?style=flat-square&maxAge=3600)

> Created with ❤️ by the Grace Bot Dev Team _aka Spooder/Gaeta_!


## Features

✅ Stable

✅ Straightforward

✅ Works with Multiple Lavalinks at a time

✅ Works with any Spotify Link

✅ Works with Huge Spotify Playlist

✅ Can set a limit on results

✅ Redis Cache System

## Installation
 
> NPM (Stable) => `npm install spotitube --save`

> Github (Dev) => `npm install botGrace/SpotiTube#master`

> Npm (Dev) => `npm install spotitube@dev`

_Installing with Redis will make the system a lot faster_

## Documentation Site

> Stable => https://SpotiTube.git.gracebot.net/

> Dev => `git clone https://github.com/BotGrace/SpotiTube#master && cd ./SpotiTube && yarn install && yarn run docs:dev`

## Getting Lavalink

Download the latest binaries from the [CI server (DEV BRANCH) (Recommended)](https://ci.fredboat.com/viewType.html?buildTypeId=Lavalink_Build&branch_Lavalink=refs%2Fheads%2Fdev&tab=buildTypeStatusDiv&guest=1) or [CI server (Stable)](https://ci.fredboat.com/viewLog.html?buildId=lastSuccessful&buildTypeId=Lavalink_Build&tab=artifacts&guest=1)

Put an [application.yml](https://github.com/freyacodes/Lavalink/blob/master/LavalinkServer/application.yml.example) file in your working directory.

Run with `java -jar Lavalink.jar`

Docker images are available on the [Docker](https://hub.docker.com/r/fredboat/lavalink/) hub.

## Getting Redis (Optional)

Download the latest version of Redis Stable [Here](https://redis.io/download) and  Read the **[Redis Quick Start (Recommended)](https://redis.io/topics/quickstart)** to get started!

## Example Usage

#### Create SpotiTube instance w/o Redis
```js
const STYT = new SpotiTube({
  spotify: {
    clientID: 'CLIENTID',
    secretKey: 'SECRETKEY'
  },
  lavalink: [{
    url: 'http://localhost:2869',
    password: 'password',
    name: 'bobross'
  }]
})
```

#### Create SpotiTube instance w/ Redis
```js
const STYT = new SpotiTube({
  spotify: {
    clientID: 'CLIENTID',
    secretKey: 'SECRETKEY'
  },
  lavalink: [{
    url: 'http://localhost:2869',
    password: 'password',
    name: 'bobross'
  }],
  redis: {
    host: "127.0.0.1",
    post: 6379,
    db: 0 // This is optional
  }
})
```

#### Convert Spotify To YT
```js
(async () => {
  const result = await STYT.convert('https://open.spotify.com/track/5nTtCOCds6I0PHMNtqelas', 15);
  console.log(result);
})();
```

## To Do List

 - [X] Redis Cache System
 - [X] Support Mutiple Lavalinks

## 🤝 Support
Contributions, issues, and feature requests are welcome! 

Give a ⭐️ if you like this project!

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

## [v1.0.6-dev](https://github.com/BotGrace/SpotiTube/compare/v1.0.3...v1.0.6-dev)

### Commits

- Added ChangeLog [`25efe7e`](https://github.com/BotGrace/SpotiTube/commit/25efe7ebb0f927c76a84067aadf59d7f028d6c3d)
- Added Redis [`a0cbf06`](https://github.com/BotGrace/SpotiTube/commit/a0cbf06d997862625b1e4493d21bd492dc4404aa)
- Full rewrite includes full playlist [`b0e7724`](https://github.com/BotGrace/SpotiTube/commit/b0e77243f338e52758a8bd470ea3847dd2f13ac0)
- fixed a cache error [`a101614`](https://github.com/BotGrace/SpotiTube/commit/a1016145ded297dbe81374bb5b83678d73fe9e05)
- Create njsscan-analysis.yml [`79863c6`](https://github.com/BotGrace/SpotiTube/commit/79863c6f2bc32d7d3f3b1bb6a56b2d1ea2509b52)
- updated readme [`3726a8b`](https://github.com/BotGrace/SpotiTube/commit/3726a8b23f4bd8f74d8579b2d83d3fa85a3036bc)
- Added proper track info [`caea1e7`](https://github.com/BotGrace/SpotiTube/commit/caea1e7d3c7ce4de97762c780942915ae9019b1b)
- added fix [`e28e555`](https://github.com/BotGrace/SpotiTube/commit/e28e55508dfa4ccdce6e1a19ba1f56b359d84493)
- added weekly downloads [`b75bff3`](https://github.com/BotGrace/SpotiTube/commit/b75bff3011cf65f3f78702848528c4c5f728cce8)
- added a filter [`acb1b49`](https://github.com/BotGrace/SpotiTube/commit/acb1b49221c85096b40ee7c1ac2e97270d0093a9)
- fixed redis error disconnect [`7a5822e`](https://github.com/BotGrace/SpotiTube/commit/7a5822edef39db73472867a37044ed0db473396a)
- added ? to program [`8b7dc74`](https://github.com/BotGrace/SpotiTube/commit/8b7dc747e47d2ce6aae457947c454015ff79279a)
- fixed tracks not listing [`c1293ce`](https://github.com/BotGrace/SpotiTube/commit/c1293cea69275be0bb017227143a7bfd0fe21b35)
- version up [`344e85e`](https://github.com/BotGrace/SpotiTube/commit/344e85ef443f6da25bad778dc9fa9cb742dfa481)
- added to do list [`47b646e`](https://github.com/BotGrace/SpotiTube/commit/47b646e4a748491bc36b811e5f0751cf257889f4)

## [v1.0.3](https://github.com/BotGrace/SpotiTube/compare/v1.0.2...v1.0.3) - 2021-08-19

### Commits

- Up [`9f192e1`](https://github.com/BotGrace/SpotiTube/commit/9f192e1bfe999bdc2b02a2564d0dfa204c684acf)
- Fix [`2024b8b`](https://github.com/BotGrace/SpotiTube/commit/2024b8b284f9c8907acecd974f7f924c30dbd8f8)
- Updated Patches [`2aff79b`](https://github.com/BotGrace/SpotiTube/commit/2aff79b9ea6e69fdd18081954ca902758b3e212c)
- moved down npm [`44c3661`](https://github.com/BotGrace/SpotiTube/commit/44c36614145af5629bb8cc608bbe1890d887f7ed)
- fix published to npm [`a84525e`](https://github.com/BotGrace/SpotiTube/commit/a84525e066b55b7f864877647c544557566baeca)
- Update package.json [`72764b2`](https://github.com/BotGrace/SpotiTube/commit/72764b2d0ad72249cf6db89d8a4203036f289d8d)
- up [`8e8db8a`](https://github.com/BotGrace/SpotiTube/commit/8e8db8aefc6937b087b99cff6577215895d7820c)
- Update gh-pages.yml [`be46f48`](https://github.com/BotGrace/SpotiTube/commit/be46f4855bdc12772262e4fdf82e0697f95179cf)

## v1.0.2 - 2021-08-19

### Merged

- e [`#1`](https://github.com/BotGrace/SpotiTube/pull/1)

### Commits

- First Commit [`94327a7`](https://github.com/BotGrace/SpotiTube/commit/94327a7ca90574a58f37dc8071fabdaef9fb3c8f)
- Test [`87d58ac`](https://github.com/BotGrace/SpotiTube/commit/87d58ac78e12d8e433f3524db1a9acd7eb31582a)
- Yeet [`877024a`](https://github.com/BotGrace/SpotiTube/commit/877024a83674025b38590930d7c952091d2b0472)
- Test [`d8223df`](https://github.com/BotGrace/SpotiTube/commit/d8223df44fe53beded35168ccbb19c83be6aae0d)
- Added Convert [`75d1421`](https://github.com/BotGrace/SpotiTube/commit/75d1421e6a1e9dc6c30d0417cf9857328ea5586d)
- Initial commit [`a7cea57`](https://github.com/BotGrace/SpotiTube/commit/a7cea5787a8063e58005fd3c7be796d863c0e756)
- created Readme [`e5cd83a`](https://github.com/BotGrace/SpotiTube/commit/e5cd83a2757a414cf234bbf342d00c064356b57a)
- All done plz work [`6112f1e`](https://github.com/BotGrace/SpotiTube/commit/6112f1e650204935c4fb4a1501d56b089334cf59)
- docusaurus! [`4684285`](https://github.com/BotGrace/SpotiTube/commit/468428561c385b1f4714939746eb4a7a8477d900)
- Fixed line 12 [`2e45c0e`](https://github.com/BotGrace/SpotiTube/commit/2e45c0e356bfe366d49d5b35bf1958329a49ee63)
- plz work [`4105770`](https://github.com/BotGrace/SpotiTube/commit/41057706ba76aba9e025a17f60cd7970e426019c)
- cleanup [`e0e027f`](https://github.com/BotGrace/SpotiTube/commit/e0e027f0c26d7bf546a5714c3e7c2349014dc1b5)
- Test Dev Build [`89fcac8`](https://github.com/BotGrace/SpotiTube/commit/89fcac83a4a58f8568dc19f1085fe2db5b844a08)

src/index.js

const Util = new require('./util');
const SpotifyWebApi = require('spotify-web-api-node');
const LavaLinkManager = require("./managers/LavaLinkManager");
const EventEmmiter = require('@tbnritzdoge/events');

/**
 * Converts Spotify To YT with the help of LavaLink
 * @class SpotiTube
 */

class SpotiTube extends EventEmmiter {
  /**
   * @description The options that SpotiTube will use to convert and link with lavalink.
   * 
   * @param {Object} options The Options Object
   * @param {Object} options.spotify The Object for Spotify
   * @param {String} options.spotify.clientID Client ID of Spotify App
   * @param {String} options.spotify.secretKey Client ID of Spotify App
   * @param {String} [options.spotify.clientAccessToken=null] Client Access Token. This will automaticlly generate
   * @param {Number} [options.spotify.clientAccessExpire=null] Client Access Token Expire Time. This will automaticlly generate
   * @param {RegExp} [options.spotify.regex=/(https?:\/\/open\.spotify\.com\/(playlist|track)\/[a-zA-Z0-9]+|spotify:(playlist|track):[a-zA-Z0-9])/g] The regex to vaildate spotify Strings
   * @param {Object[]} options.lavalink The Object for LavaLink
   * @param {String} options.lavalink.url The lavalink url w/ http:// or https://
   * @param {String} options.lavalink.password Lavalink password 
   * @param {String} [options.lavalink.name] Lavalink node name
   * @param {Object=} options.redis The Object for Redis (To use redis put host & port)
   * @param {String=} options.redis.host The ip of redis (Redis default is 127.0.0.1)
   * @param {String=} options.redis.password The password of redis if one exist
   * @param {Number=} [options.redis.port=6379] The port of redis defaults to 6379
   * @param {Number=} options.redis.db The db to use on redis
   * @returns {Object}
   * 
   * @example
   * const STYT = new SpotiTube({
   *    spotify: {
   *      clientID: 'CLIENTID',
   *      secretKey: 'SECRETKEY'
   *    },
   *    lavalink: [{
   *      url: 'http://localhost:2869',
   *      password: 'password'
   *    }],
   *   redis: {
   *      host: "127.0.0.1",
   *      post: 6379,
   *      db: null
   *   }
   * })
   */
  constructor(options = {}) {
    super()
    this.options = Util.mergeDefault({
      spotify: {
        clientID: null,
        secretKey: null,
        clientAccessToken: null,
        clientAccessExpire: null,
        regex: /(https?:\/\/open\.spotify\.com\/(playlist|track)\/[a-zA-Z0-9]+|spotify:(playlist|track):[a-zA-Z0-9])/g
      },
      lavalink: [{
        url: null,
        password: null
      }],
      redis: {
        host: null,
        password: null,
        post: 6379,
        db: null
      }
    }, options)

    /**
     * The list of supported spotify links we support
     * @type {String[]}
     */
    this.supportedTypes = ['playlist', 'track'];

    // Check for correct strings
    if (!this.options.spotify.clientID || !this.options.spotify.secretKey) throw new Error('Missing Spotify Client ID Or Spotify Secrect Key');

    /**
     * The Manager that stores all the lavalinks
     * @type {LavaLinkManager}
     */
    this.lavalinks = new LavaLinkManager(this.options.lavalink);

    // Load Creds
    /**
     * The Spotify Web API
     * @type {SpotifyWebApi}
     */
    this.spotifySearch = new SpotifyWebApi({
      clientId: this.options.spotify.clientID,
      clientSecret: this.options.spotify.secretKey
    });

    // Load Redis
    if (this.options.redis.host && this.options.redis.port) {

      this.redis = require("redis").createClient(this.options.redis);

      // Redis Error Event Reconnect
      this.redis.on("error", err => {
        if (err?.toString()?.includes('ECONNRESET')) {
          setTimeout(() => {
            this.redis = this.redis.createClient(this.options.redis);
          }, 15000);
        }
      });

      // Redis Error Event
      this.redis.on("error", (args) => this.emit('error', args || "Redis Error"));

      // Redis Ready Event
      this.redis.on("ready", (args) => this.emit('debug', args || "Redis Ready"));

      // Redis Connect Event
      this.redis.on("connect", (args) => this.emit('debug', args || "Redis Connected"));

      // Redis Reconnecting Event
      this.redis.on("reconnecting", (args) => this.emit('debug', args || "Redis Reconnecting"));
    } else {
      this.redis = null;
      this.emit("debug", "No Redis Host & Port. Redis will not be used.")
    }

    // Load Any Unknown Errors
    process.on('uncaughtException', (args) => this.emit('error', args));

    // Retrieve an access token.
    if (this.options.spotify.clientAccessToken && this.options.spotify.clientAccessExpire) this.initCreds({force: true, access_token: this.options.spotify.clientAccessToken, expires_in: this.options.spotify.clientAccessExpire})
    // else this.spotifySearch.clientCredentialsGrant().then(res => this.initCreds(res.body)).catch(e => console.log(e))

  }

  /**
   * @description Debug related events
   * This will include Redis Events
   * 
   * @event SpotiTube#debug
   * @param {String} info The debug info
   * @memberof SpotiTube
   * 
   * @example
   * SpotiTube.on("debug", console.log);
   */

  /**
   * @description When the process has an error event.
   * This will include Redis Error Event
   * 
   * @event SpotiTube#error
   * @param {Error} error The error object
   * @memberof SpotiTube
   * 
   * @example
   * SpotiTube.on("error", console.log);
   */

  /**
   * @description Loads/Creates/Checks the auth token to the Spotify Web API.
   * 
   * @param {Object} settings The object of the data.
   * @param {String} settings.access_token The access token to use.
   * @param {String} [settings.token_type=Bearer] The token type.
   * @param {Number} settings.expires_in The time of when it will expire in seconds.
   * @returns {Object}
   * @memberof SpotiTube
   */

  initCreds (settings = {}) {
    settings = Util.mergeDefault({
      access_token: null,
      token_type: "Bearer",
      expires_in: null,
      force: false
    }, settings)
    if ((!this.options.spotify.clientAccessExpire || !this.options.spotify.clientAccessExpire) && (!settings.access_token || !settings.expires_in) && !settings.force) {
      this.emit("debug", "There is no login detect and none were sent with function. Creating login and restarting function")
      return this.spotifySearch.clientCredentialsGrant().then(res => this.initCreds(res.body)).catch(e => this.emit("error", e));
    } else if ((new Date(new Date().getTime() + (1000 * 120)) >= this.options.spotify.clientAccessExpire) && (!settings.access_token || !settings.expires_in) && !settings.force) {
      this.emit("debug", "Current access token is expires or is within the 2 min mark. Creating new login and restarting function");
      return this.spotifySearch.clientCredentialsGrant().then(res => this.initCreds(res.body)).catch(e => this.emit("error", e));
    } else if (settings.access_token || settings.expires_in) {
      this.spotifySearch.setAccessToken(settings.access_token)
      this.options.spotify.clientAccessToken = settings.access_token;
      this.options.spotify.clientAccessExpire = new Date(new Date().getTime() + (1000 * settings.expires_in));
      this.emit("debug", `Added Access token & Expire Time (${this.options.spotify.clientAccessExpire.getTime()})`);
      return this.spotifySearch
    } else {
      return this.spotifySearch
    }
  }
  
  /**
   * @description Validate that the URL is a Spotify URL
   * 
   * @param {Url} url The URL of the Spotify Track Or Playlist
   * @returns {Boolean}
   * @memberof SpotiTube
   * 
   * @example
   * (async () => {
   *    const result = await STYT.validateURL('https://open.spotify.com/track/5nTtCOCds6I0PHMNtqelas');
   *    console.log(result)
   * })();
   * 
   * @example
   * (async () => {
   *    const result = await STYT.validateURL('https://open.spotify.com/playlist/5B30UzmQQ6exwcZPwA8tbF?si=9df28e96ebf34267');
   *    console.log(result)
   * })();
   * 
   * @example
   * (async () => {
   *    const result = await STYT.validateURL('spotify:track:4aDSp2TuP7OSPvN9wrwcs5');
   *    console.log(result)
   * })();
   * 
   * @example
   * (async () => {
   *    const result = await STYT.validateURL('spotify:playlist:5B30UzmQQ6exwcZPwA8tbF');
   *    console.log(result);
   * })();
   */
  async validateURL (url) {
    if (!url) throw new Error('You did not specify the URL of Spotify!');
    if (typeof url !== 'string') return false;
    if (new RegExp(this.options.spotify.regex).test(url)) {
      let parsedURL = {}
      try {
        parsedURL = require('spotify-uri')?.parse(url) || null;
        if (!this.supportedTypes.includes(parsedURL?.type)) return false;
        if (!parsedURL) return false;
        return true;
      } catch (e) {
        return false;
      }
    } else return false;
  }

  /**
   * @description Get info like type, name, and authors on the given url.
   * 
   * @param {URL} url The url you want to check.
   * @returns {Object}
   * @memberof SpotiTube
   * 
   * @example
   * (async () => {
   *    const result = await STYT.getInfo('https://open.spotify.com/track/5nTtCOCds6I0PHMNtqelas');
   *    console.log(result);
   * })();
   */
  async getInfo(url) {
    if (!url) throw new Error('You did not specify the URL of Spotify!');
    if (!this.validateURL(url)) throw new Error('Url is not a match to the regex!') ;
    try {
      let data = await require('spotify-url-info')?.getData(url) || null;
      this.emit("debug", `${url} = ${data.type} from getInfo`)
      return data;
    } catch (error) {
      this.emit("error", error);
      return null
    }
  }

  /**
   * @description Gets the data off a playlist.
   * 
   * @param {String} id The url you want to check.
   * @param {Object} options Addtional options to add
   * @param {Number} [options.offset=0] How many tracks to skip
   * @param {Number} [options.limit=100] How many tracks to skip (100 Max with 0 Min)
   * @returns {Object}
   * @private
   * @memberof SpotiTube
   */

  async getPlaylist (id, options = {}) {
    if (!id) throw new Error("Missing playlist ID")
    options = Util.mergeDefault({
      offset: 0,
      limit: 100
    }, options)
    await this.initCreds(); // Make sure creds are correct
    let data = await this.spotifySearch.getPlaylistTracks(id, {offset: options.offset, limit: options.limit}) || null;
    return data?.body || null
  }

  /**
   * @description Checks Redis Cache.
   * 
   * @param {String} key The url you want to check.
   * @returns {Object}
   * @private
   * @memberof SpotiTube
   */

  async getRedisCache (key) {
    if (!key) throw new Error("Reddis Key");
    if (!this.redis) return null;
    return new Promise(async (resolve, reject) => {
      this.redis.get(key, async (error, reply) => {
        if (error) {
          this.emit("error", error);
          return resolve(null)
        }
        else {
          if (reply) {
            try {
              if (JSON.parse(reply)) {
                return resolve(JSON.parse(reply))
              } else {
                return resolve(reply)
              }
            } catch (e) {
              return resolve(reply)
            }
          } else return resolve(null)
        }
      })
    })
  }

  /**
   * @description Set Redis Cache.
   * 
   * @param {String} key The url you want to check.
   * @param {*} data The data to be cached.
   * @returns {*}
   * @private
   * @memberof SpotiTube
   */

   async setRedisCache (key, data) {
    if (!key) throw new Error("Reddis Key");
    if (!data) throw new Error("Needs Value");
    if (!this.redis) return null;

    try {
      if (JSON.stringify(data)) {
        data = JSON.stringify(data)
      }
    } catch (e) {
      console.log(e)
    }

    return new Promise(async (resolve, reject) => {
      this.redis.set(key, data, async (error) =>{
        if (error) {
          this.emit("error", error)
          return resolve(false)
        } else {
          this.emit("debug", `Key ${key} was set to ${data}`);
          return resolve(true)
        }
      });
    })

  }

  /**
   * @description Converts the spotify url(s) to a youtube result.
   * The Longer the playlist on spotify the longer it will take
   * 
   * @param {URL} url The url you want to convert.
   * @param {Number} [limit=20] Limit how many songs we should convert. Use Infinity to allow the entire playlist, but the bigger the playlist the longer it will take to convert
   * @param {Boolean} [failedLimit=true] Let failed song searches include in the overall limit checks.
   * @returns {Object}
   * @memberof SpotiTube
   * 
   * @example
   * (async () => {
   *     const result = await STYT.convert('https://open.spotify.com/track/5nTtCOCds6I0PHMNtqelas', 200, true);
   *     console.log(result);
   * })();
   * 
   * @example
   * (async () => {
   *     const result = await STYT.convert('https://open.spotify.com/track/5nTtCOCds6I0PHMNtqelas', Infinity, true); // Infinity will allow you to get all results
   *     console.log(result);
   * })();
   * 
   * @example
   * (async () => {
   *     const result = await STYT.convert('spotify:playlist:5SRG4654V94q7ao2X1hUU1', Infinity, true); // Infinity will allow you to get all results
   *     console.log(result);
   * })();
   * 
   * @example
   * (async () => {
   *     const result = await STYT.convert('spotify:track:4aDSp2TuP7OSPvN9wrwcs5', Infinity, true); // Infinity will allow you to get all results
   *     console.log(result);
   * })();
   */

  async convert(url, limit = 20, failedLimit = true) {
    if (!url) throw new Error('You did not specify the URL of Spotify!');
    if (!this.validateURL(url)) throw new Error('Url is not a match to the regex!');
    var start = new Date();
    await this.initCreds() // Make sure creds are correct
    if (!limit || limit < 1) limit = 20
    if (failedLimit !== false && failedLimit !== true) failedLimit = false    
    this.emit("debug", `Getting ${url} with limit of ${limit} with ${failedLimit ? 'failed searches being included with limit' : 'failed searches not being included with limit'}`)

    let getInfo = await this.getInfo(url);

    if (!getInfo) {
      this.emit("debug", `${url} was not found`)
      return null
    }

    this.emit("debug", `${url} = ${getInfo.type} in convert`)

    if (!this.supportedTypes.includes(getInfo.type)) throw new Error(`${getInfo.type} is not a support format only ${this.supportedTypes.join()}`);

    const node = await this.lavalinks.getBestNode();
    this.emit("debug", `Using node ${node.name}`);

    if (getInfo.type === "track") {
      // track
      this.emit("debug", `${url} is an ${getInfo.type} so limits will not work on this`)
      let result;
      try {
        if (this.redis) {
          let check = await this.getRedisCache(`${getInfo.uri || 'spotify:' + getInfo?.external_urls?.spotify}`);
          if (!check) {
            result = await node.search(`${getInfo.name} ${getInfo.artists.map(x => x.name).join(' ')}`);
            await this.setRedisCache(`${getInfo?.uri || 'spotify:' + getInfo?.external_urls?.spotify}`, result);
            this.emit("debug", `${getInfo?.external_urls?.spotify || getInfo?.uri} => ${result.uri} (Not From Cache)`)
          } else {
            result = check;
            this.emit("debug", `${getInfo?.external_urls?.spotify || getInfo?.uri} => ${result.uri} (From Cache)`)
          }
        } else {
          const node = await this.lavalinks.getBestNode();
          this.emit("debug", `Using node ${node.name}`)
          result = await node.search(`${getInfo.name} ${getInfo.artists?.map(x => x.name).join(' ')}`);
          this.emit("debug", `${getInfo?.external_urls?.spotify || getInfo?.uri} => ${result.uri}  (Not From Cache. Redis not being used)`)
        }
      } catch (error) {
        result = null;
        this.emit("debug", `${url} was not found`)
        this.emit("error", error)
      }
      return {
        songs: {
          failed: !result ? [`${getInfo?.external_urls?.spotify || getInfo.uri}`] : [],
          completed: result ? [{
            url: result.uri,
            info: result,
            track: result.uri.track,
            spotify: {
              key: getInfo?.uri,
              url: getInfo?.external_urls?.spotify,
              name: getInfo?.name,
              artists: getInfo?.artists || []
            }
          }] : []
        },
        info: getInfo,
        limit: limit,
        converted: {
          failed: !result ? 1 : 0,
          completed: result ? 1 : 0,
          total: 1
        },
        time: {
          start: start,
          end: new Date(),
          executionMS: new Date() - start,
          executionSec: (new Date() - start) / 1000
        }
      }
    } else if (getInfo.type === "playlist") {
      // playlist
      let data = await this.getPlaylist(getInfo.id);
      let tracks = [];
      for (const song of data.items) {
        if (tracks.length >= limit) break;
        else tracks.push(song)
      }
      let next = data?.next ? true : false;
      while (next === true && (tracks.length < limit)) {
        let data2 = await this.getPlaylist(getInfo.id, {offset: tracks.length, limit: (data.total - tracks.length) >= 100 ? 100 : data.total - tracks.length })
        for (const song of data2.items) {
          if (tracks.length >= limit) break;
          tracks.push(song)
        }
        if (tracks.length >= limit) next = false;
        else next = data2?.next ? true : false        
      }
      this.emit("debug", `Gathered a total of ${tracks.length} with a limitter set to ${limit}`)
      var songs = [];
      var failed = [];
      for (let song of tracks) {
        song = song.track
        if ((failedLimit ? failed.length + songs.length : songs.length) >= limit) break;
        this.emit("debug", `${url} Current searches: ${(failedLimit ? failed.length + songs.length : songs.length)} w/ limit of ${limit}.`)
        let result;
        if (!song?.uri || !song?.external_urls?.spotify) {
          failed.push(song?.external_urls?.spotify || song?.uri || song?.name || "Unknown song")
        } else {
          try {
            if (this.redis) {
              let check = await this.getRedisCache(`${song.uri || 'spotify:' + song?.external_urls?.spotify}`);
              if (!check) {
                result = await node.search(`${song.name} ${song.artists.map(x => x.name).join(' ')}`);
                await this.setRedisCache(`${song?.uri || 'spotify:' + song?.external_urls?.spotify}`, result);
                this.emit("debug", `${song?.external_urls?.spotify || song?.uri} => ${result.uri} (Not From Cache)`)
              } else {
                result = check;
                this.emit("debug", `${song?.external_urls?.spotify || song?.uri} => ${result.uri} (From Cache)`)
              }
            } else {
              result = await node.search(`${song.name} ${song.artists.map(x => x.name).join(' ')}`);
              this.emit("debug", `${song?.external_urls?.spotify || song?.uri} => ${result.uri}  (Not From Cache. Redis not being used)`)
            }
          } catch (error) {
            result = null;
            this.emit("debug", `${song?.external_urls?.spotify || song?.uri || song?.name || "Unknown song"} was not found`)
            this.emit("error", error)
          }
          if (!result) failed.push(song?.external_urls?.spotify || song?.uri || song?.name || "Unknown song")
          else songs.push({
            url: result.uri,
            info: result,
            track: result.uri.track,
            spotify: {
              key: song?.uri,
              url: song?.external_urls?.spotify,
              name: song?.name,
              artists: song?.artists || []
            }
          })
        }
      }
      this.emit("debug", `${getInfo?.external_urls?.spotify || getInfo.uri} was converted to ${songs.length} with ${failed?.length} with limit ${limit}`)
      return {
        songs: {
          failed: failed,
          completed: songs
        },
        info: {...getInfo, tracks: tracks.filter(g => !g?.track?.uri)?.map(g => {
          return `${g?.track?.uri || 'spotify:' + g?.track?.external_urls?.spotify}`
        }) || []},
        limit: limit,
        converted: {
          failed: failed?.length || 0,
          completed: songs?.length || 0,
          total: (failed?.length + songs?.length) || 0
        },
        time: {
          start: start,
          end: new Date(),
          executionMS: new Date() - start,
          executionSec: (new Date() - start) / 1000
        }
      }
    } else throw new Error(`${getInfo.type} is not a support format only ${this.supportedTypes.join()}`)
  }

  /**
   * @description Search on lavalink.
   * @deprecated Use {@link LavaLink.search} instead
   * 
   * @param {String} query The search query to send to lavalink.
   * @returns {Object}
   * 
   * @example
   * (async () => {
   *    const result = await STYT.searchLavaLink('say something');
   *    console.log(result);
   * })();
   */

  searchLavaLink () {
    process.emitWarning('This function event is deprecated. Use lavalink search instead', 'DeprecationWarning');
  }

}

module.exports = SpotiTube
